### PR TITLE
feat: Seed key for rand primitive

### LINKED ISSUE #123: [FEATURE] Seed key for random array creation primitives
# Feature Request

## Describe the Feature
Add seed key for the random array creation primitives.

## Motivation
To be able to reproduce experiments we need to able to set seed of the random array creation primitives. However, currently this behavior is not supported.

### PR DESCRIPTION
**Pull request by aturker-synnada**
## Description

Closes #123 

## What is Changed

Include the changes introduced in this PR.

- Random seed values added to datastore.
- FlatGraph keep track of random seed keys.
- PM updates seed values in every evaluation.

## Checklist:

- [x] Tests that cover the code added.
- [x] Corresponding changes documented.
- [x] All tests passed.
- [x] The code linted and styled (pre-commit run --all-files has passed).

### CODE REVIEW

File: mithril/framework/physical/data_store.py [hunk 1]
```diff
@@ -82,6 +87,10 @@ def all_static_keys(self) -> set[str]:
     def unused_keys(self) -> set[str]:
         return self._unused_keys
 
+    @property
+    def random_seed_keys(self) -> set[str]:
```
Comment (kberat-synnada) [inline mithril/framework/physical/data_store.py:91]: not used

File: mithril/framework/physical/data_store.py [hunk 2]
```diff
@@ -61,6 +61,7 @@ def __init__(
         # Final tensor values of data store.
         self.data_values: DataEvalType[DataType] = dict()
         self.constraint_solver: ConstraintSolver = deepcopy(solver, memo=memo)
+        self._random_seed_values: dict[str, int] = dict()
```
Comment (kberat-synnada) [inline mithril/framework/physical/data_store.py:64]: we can say self._random_seeds since it is a dict

File: mithril/framework/logical/base.py [hunk 3]
```diff
@@ -145,6 +145,7 @@ def __init__(self, name: str | None = None, enforce_jit: bool = True) -> None:
         self.constraint_solver: ConstraintSolver = ConstraintSolver()
         self.safe_shapes: dict[str, ShapeTemplateType] = {}
         self.is_frozen = False
+        self._random_keys: set[str] = set()
```
Comment (kberat-synnada) [inline mithril/framework/logical/base.py:148]: move to Primitive

File: mithril/framework/physical/data_store.py [hunk 4]
```diff
@@ -396,3 +411,21 @@ def infer_static_keys(self) -> Updates:
...
         return updates
+
+    def set_random_seed_keys(self, seed_keys: set[str]):
+        for key in seed_keys:
+            if self.all_data[key].value == TBD:
+                self._random_seed_values[key] = 0
+            else:
+                self._random_seed_values[key] = self.all_data[key].value
+
+    def set_random_seed_values(self, seed_values: dict[str, int]):
```
Comment (kberat-synnada) [inline mithril/framework/physical/data_store.py:422]: could we make inputs keyworded
def set_random_seed_values(self, **kwargs):

**Review by kberat-synnada — APPROVED** [review]

**Review by norhan-synnada — APPROVED** [review]

**Pull request merged**